### PR TITLE
Change reconnect logic so sessions are revived upon reconnect

### PR DESCRIPTION
**Pull request by austinzheng**
~WIP pending integration with Cocoa changes and manual verification.~ It works

Changes:
- When the reachability system detects network connectivity, the sync manager's `reconnect()` method is now called, rather than directly calling the client cancel delay method.
- The sync manager's `reconnect()` method now asks each session to revive itself if needed.
- If a session is revived in the 'waiting for token' state, the session will now run the login callback. For the Cocoa binding, this simply means that any existing refresh handle will be cancelled and a new one will be created, causing a login request to be sent immediately.

### CODE REVIEW

**Comment by austinzheng** [comment]
I changed the way this works because `revive_if_needed()` wasn't what we wanted.

`revive_if_needed()` is intended for taking a session which was inactive or slated to be destroyed and indicating that someone does want to use it further.

The new `handle_reconnect()` method has a different meaning. If the session is already active or intended to be active, it'll trigger a token refresh. Otherwise, it does nothing.

File: src/sync/sync_session.cpp [hunk 1]
```diff
@@ -184,7 +187,15 @@ struct sync_session_states::WaitingForAccessToken : public SyncSession::State {
     bool revive_if_needed(std::unique_lock<std::mutex>&, SyncSession& session) const override
     {
         session.m_deferred_close = false;
-        return false;
+        return true;
```
Comment (bdash) [inline src/sync/sync_session.cpp]: Why is this necessary?

Comment (austinzheng) [inline src/sync/sync_session.cpp]: That should be `false`. I thought I changed it back...
